### ISSUE TITLE
Int boxing bug

### ISSUE THREAD

**Issue by japgolly** [scala-js/scala-js#2068]
I started working on official Scala.JS support for Scalaz.
One of the tests `MapTest` fails thus:

```
[info] ! .==>> map:mapKeys sound: Exception raised on property evaluation.
[info] > ARG_0: Bin(1740834887,1268128331,Bin(-1421249690,1876982179,Tip,Tip),Bin(2147483647,-2147483648,Tip,Tip))
[info] > Exception: scala.scalajs.runtime.UndefinedBehaviorError: An undefined behavior was detected: 0 is not an instance of java.lang.Integer
```

It seems that the `insert` method of `Map` (aka `==>>`) gets an `Order[java.lang.Integer]` and tries to feed it an `Int`. This only happens under Scala.JS so I think it's a Scala.JS bug.

You can reproduce by checking out my fork: https://github.com/japgolly/scalaz/tree/scalajs
Ensure you're on the `scalajs` branch, then run `sbt 'testsJS/test-only scalaz.MapTest'`.
(If it passes, run it again; the default settings only perform 33 tests per property.)

**Comment by japgolly** [comment]
(Also the tests don't run in parallel like my other projects with μTest. I assumed it was the Scala.JS SBT plugin that forked multiple instances of node/phantomjs. Sequentially is super slow - where do I look next to fix this? Settings? ScalaCheck? SJS?)

**Comment by andreaTP** [comment]
Can you please point out some lines number to look at?
An ```Ordering``` is created only when a Sorted data structure is used and it is not a general case...

Probably if you can identify the exact data structure it will be really helpful!

**Comment by japgolly** [comment]
Sure. I already pointed out the data structure, test and method though, is that not what you mean? What I say the `insert` method of `Map` and `MapTest`, those are files in the Scalaz repo, I don't mean anything about the `Map`  in stdlib. Maybe I wasn't clear about that.

Let me provide some links. The failing test is:
https://github.com/japgolly/scalaz/blob/scalajs/tests/shared/src/test/scala/scalaz/MapTest.scala#L591-L599

The method that crashes:
https://github.com/japgolly/scalaz/blob/scalajs/core/shared/src/main/scala/scalaz/Map.scala#L32-L46

Specifically, I think this is the exact line:
https://github.com/japgolly/scalaz/blob/scalajs/core/shared/src/main/scala/scalaz/Map.scala#L38

**Comment by sjrd** [comment]
I will have a look at this next week. This week I'm completely overwhelmed with academic stuff.

**Comment by japgolly** [comment]
@sjrd No worries. Thanks for letting me know :)

**Comment by sjrd** [comment]
I think this is a duplicate of #1984, which is fixed in master. Guessing that from the `%` in the test.

**Comment by sjrd** [comment]
I cannot reproduce the bug on your branch. I have tried running it ~15 times and it didn't fail.

If you can still reproduce it with 0.6.5, could you try with a locally-published 0.6.6-SNAPSHOT?

**Comment by japgolly** [comment]
That's weird, I ran it another 10 times just now and it only passed once.

I also tried it with a local 0.6.6-SNAPSHOT and it continually passes now. It must've been the `%` bug like you suggested.

**Comment by sjrd** [comment]
OK, thanks for trying. I'll close this as a duplicate of #1984, then.
